File: src/cleaner/ChannelCleaner.java
package cleaner;

import java.time.Instant;
import java.time.LocalTime;
import java.util.Calendar;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;
import java.util.stream.Collectors;
import discord4j.core.object.entity.Message;
import discord4j.core.object.entity.TextChannel;
import discord4j.core.object.util.Snowflake;
import reactor.core.publisher.Flux;

public class ChannelCleaner {
	
	private TextChannel channel;
	private Calendar cal;
	
	
	public ChannelCleaner(TextChannel channel) {
		this.channel = channel;
		cal = Calendar.getInstance();
		int day = cal.get(Calendar.DAY_OF_MONTH);
		cal.set(Calendar.DAY_OF_MONTH, day+1);
		cal.set(Calendar.HOUR_OF_DAY, 00);
		cal.set(Calendar.MINUTE, 01);
		cal.set(Calendar.SECOND, 00);
		
		System.out.println(LocalTime.now() + " Cleaning job started for channel: " 
				+ channel.getGuild().block().getName() + "/" + channel.getName() + "\n"
				+ "First clean scheduled at " + this.cal.getTime()
				);
		
		Timer timer = new Timer();
		timer.schedule(new CleanTask(timer), cal.getTime());
	}
	
	private void clean() {
		System.out.println(LocalTime.now() + " Cleaning channel " + channel.getGuild().block().getName() + "/" + channel.getName());
		//max 2 weeks old messages
		Instant i = Instant.ofEpochSecond(Instant.now().getEpochSecond()-1209600);
		List<Snowflake> messages = channel.getMessagesBefore(Snowflake.of(Instant.now())).take(100)
			.filter(msg -> msg.getTimestamp().isAfter(i))
			.filter(msg -> !msg.isPinned())
			.map(Message::getId)
			.collect(Collectors.toList())
			.block();
		
		while(messages.size() > 0) {
			//bulk delete messages
			channel.bulkDelete(Flux.fromIterable(messages)).subscribe();
			messages = channel.getMessagesBefore(Snowflake.of(Instant.now())).take(100)
					.filter(msg -> msg.getTimestamp().isAfter(i))
					.filter(msg -> !msg.isPinned())
					.map(Message::getId)
					.collect(Collectors.toList())
					.block();
					
		}
		
		System.out.println("finished cleaning in channel: " + channel.getGuild().block().getName() + "/" + channel.getName());
	}
	
	public void printCleaningTime() {
		System.out.println(channel.getGuild().block().getName() + "/" + channel.getName() + " cleaning time is: " + cal.getTime().toString());
	}
	
	public void cleanNow() {
		clean();
	}
	
	class CleanTask extends TimerTask {
		private Timer timer;
		
		public CleanTask(Timer timer) {
			this.timer = timer;
		}
		
		@Override
		public void run() {
			clean();
			timer.cancel();
			int day = cal.get(Calendar.DAY_OF_MONTH);
			cal.set(Calendar.DAY_OF_MONTH, day+1);
			cal.set(Calendar.HOUR_OF_DAY, 00);
			cal.set(Calendar.MINUTE, 01);
			cal.set(Calendar.SECOND, 00);
			Timer newt = new Timer();
			newt.schedule(new CleanTask(newt), cal.getTime());
		}
	}
}

File: src/main/kotlin/cleaner/ChannelCleaner.kt
package cleaner

import discord4j.common.util.Snowflake
import discord4j.core.`object`.entity.Message
import discord4j.core.`object`.entity.channel.GuildMessageChannel
import raid.RaidMessageHandler
import java.time.Instant
import java.time.LocalTime
import java.util.*

class ChannelCleaner(private val channel: GuildMessageChannel,
                        private val handler: RaidMessageHandler) {

    private val twoWeeks: Instant = Instant.ofEpochSecond(Instant.now().epochSecond - 1209600)
    private val timer: Timer = Timer()
    private var schedule: Date

    init {
        println("Cleaner task started for channel ${channel.name}")
        schedule = getScheduleTime()
        timer.schedule(CleanTask(), schedule)
        println("First clean scheduled at $schedule")
    }

    private fun getScheduleTime(): Date {
        return Calendar.getInstance()
                .also {
                    val day = it.get(Calendar.DAY_OF_MONTH)
                    it.set(Calendar.DAY_OF_MONTH, day+1)
                    it.set(Calendar.HOUR_OF_DAY, 0)
                    it.set(Calendar.MINUTE, 1)
                    it.set(Calendar.SECOND, 0)
                }.time
    }

    private fun clean() {
        println("${LocalTime.now()} Cleaning started in ${channel.name}")
        // count things
        var count = channel.getMessagesBefore(Snowflake.of(Instant.now())).take(100)
                .filter { msg -> msg.timestamp.isAfter(twoWeeks) }
                .filter { msg -> !msg.isPinned }
                .map(Message::getId)
                .count()
                .block()
        while(count!! > 0) {
            removeBatch()
            count = channel.getMessagesBefore(Snowflake.of(Instant.now())).take(100)
                    .filter { msg -> msg.timestamp.isAfter(twoWeeks) }
                    .filter { msg -> !msg.isPinned }
                    .map(Message::getId)
                    .count()
                    .block()
        }
        println("${LocalTime.now()} Cleaning complete in ${channel.name}")
    }

    private fun removeBatch() {
        val messages = channel.getMessagesBefore(Snowflake.of(Instant.now())).take(100)
                .filter{ msg -> msg.timestamp.isAfter(twoWeeks)}
                .filter { msg -> !msg.isPinned }
                .map (Message::getId)
                .transform(channel::bulkDelete)

        val last = messages.blockLast()
        last?.let {
            channel.getMessageById(it).block()?.delete()?.block()
        }
    }

    fun cleanNow() {
        println("${LocalTime.now()} Cleaning started in ${channel.name}")
        clean()
    }

    fun printCleanTime() {
        println("${channel.name} cleaning scheduled at $schedule")
    }

    inner class CleanTask: TimerTask() {
        override fun run() {
            clean()
            println("${LocalTime.now()} Clearing raids...")
            handler.clearRaids()
            Timer().apply {
                //val time = Date(Calendar.getInstance().timeInMillis + (10 * 3600))
                schedule = getScheduleTime()
                println("New schedule in channel ${channel.name} $schedule")
                this.schedule(CleanTask(), schedule)
            }
        }

    }
}